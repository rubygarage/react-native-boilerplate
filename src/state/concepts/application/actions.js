import * as types from './types';

export const setOnboardingCompletion = (isCompleted) => ({
  type: types.SET_ONBOARDING_COMPLETION,
  isCompleted,
});

export const startListenConnectionState = () => ({
  type: types.START_LISTEN_CONNECTION_STATE,
});

export const stopListenConnectionState = () => ({
  type: types.STOP_LISTEN_CONNECTION_STATE,
});

export const setConnectionState = (isConnected) => ({
  type: types.SET_CONNECTION_STATE,
  isConnected,
});
