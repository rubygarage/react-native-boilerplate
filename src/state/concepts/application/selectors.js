import { path } from 'ramda';

export const isOnboardingCompletedSelector = path(['application', 'onboarding']);

export const isConnectedSelector = path(['application', 'connection']);
