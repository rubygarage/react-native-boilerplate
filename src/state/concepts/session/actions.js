import * as types from './types';

export const setTokens = (tokens) => ({
  type: types.SET_TOKENS,
  tokens,
});

export const removeTokens = () => ({
  type: types.SET_TOKENS,
  tokens: null,
});

export const setUserId = (id) => ({
  type: types.SET_USER_ID,
  id,
});
