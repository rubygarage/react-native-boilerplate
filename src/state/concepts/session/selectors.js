import { createSelector } from 'reselect';
import build from 'redux-object';
import { path, isNil, complement, prop } from 'ramda';

const dataSelector = prop('data');

export const tokensSelector = path(['session', 'tokens']);

export const isAuthorizedSelector = createSelector(tokensSelector, complement(isNil));

export const currentUserIdSelector = path(['session', 'currentUserId']);

export const currentUserSelector = createSelector(
  currentUserIdSelector,
  dataSelector,
  (id, data) => (data.clientProfile && data.clientProfile[id] ? build(data, 'clientProfile', id) : null),
);
