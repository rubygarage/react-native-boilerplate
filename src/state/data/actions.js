import * as types from './types';

export const dataApiRequest = ({ endpoint }) => ({
  type: types.DATA_API_REQUEST,
  endpoint,
});

export const dataApiSuccess = ({ endpoint, response }) => ({
  type: types.DATA_API_SUCCESS,
  endpoint,
  response,
});

export const dataApiFailure = ({ endpoint }) => ({
  type: types.DATA_API_FAILURE,
  endpoint,
});

export const dataDelete = ({ kind, ids }) => ({
  type: types.DATA_DELETE,
  kind,
  ids,
});
