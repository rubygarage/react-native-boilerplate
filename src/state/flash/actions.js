import { ALERT_TYPES } from 'constants/alert';
import * as types from './types';

export const showFlash = ({
  id,
  lifetime,
  priority,
  isSingleShot = false,
  flashProps,
}) => ({
  type: types.SHOW_FLASH,
  flash: {
    id,
    lifetime,
    priority,
    isSingleShot,
    flashProps,
  },
});

export const showErrorFlash = (action) => ({
  type: types.SHOW_FLASH,
  flash: {
    isSingleShot: false,
    flashProps: {
      title: { id: 'shared.oops' },
      type: ALERT_TYPES.error,
      action,
      actionTextId: 'shared.tryAgain',
      isCloseable: true,
    },
  },
});

export const addFlash = (flash) => ({
  type: types.ADD_FLASH,
  flash,
});

export const hideFlash = (id) => ({
  type: types.HIDE_FLASH,
  id,
});

export const removeFlash = (id, isSingleShot) => ({
  type: types.REMOVE_FLASH,
  id,
  isSingleShot,
});
