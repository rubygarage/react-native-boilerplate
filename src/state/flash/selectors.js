import { createSelector } from 'reselect';
import {
  prop, path, sort, ascend, pipe, values,
} from 'ramda';

const listSelector = path(['flash', 'list']);

export const flashSelector = createSelector(
  listSelector,
  (_, id) => id, (flashes, id) => prop(id, flashes),
);

const flashValuesSelector = pipe(listSelector, values);

export const flashesSelector = createSelector(
  flashValuesSelector,
  sort(ascend(prop('priority'))),
);

export const viewedIdsSelector = path(['flash', 'viewedIds']);
