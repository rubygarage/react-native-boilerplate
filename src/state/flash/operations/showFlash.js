import { createLogic } from 'redux-logic';
import { contains } from 'ramda';

import { SHOW_FLASH } from '../types';
import { viewedIdsSelector } from '../selectors';
import { addFlash } from '../actions';

const showFlash = createLogic({
  type: SHOW_FLASH,
  latest: true,

  process({ action: { flash }, getState }, dispatch, done) {
    const viewedIds = viewedIdsSelector(getState());
    const { id } = flash;

    if (!contains(id, viewedIds)) {
      dispatch(addFlash(flash));
    }
    done();
  },
});

export default showFlash;
