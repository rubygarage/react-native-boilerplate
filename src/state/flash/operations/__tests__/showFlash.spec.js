import showFlash from '../showFlash';
import { addFlash } from '../../actions';
import { viewedIdsSelector } from '../../selectors';

jest.mock('../../selectors', () => ({
  viewedIdsSelector: jest.fn(() => []),
}));

describe('showFlash', () => {
  const flash = { id: '300' };
  const action = { flash };
  const getState = jest.fn();
  const dispatch = jest.fn();
  const done = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('has valid attributes', () => {
    expect(showFlash).toMatchSnapshot();
  });

  it('dispatches "addFlash" action', () => {
    showFlash.process({ action, getState }, dispatch, done);
    expect(dispatch).toHaveBeenCalledWith(addFlash(flash));
    expect(done).toHaveBeenCalled();
  });

  it('does not dispatch "addFlash" action', () => {
    viewedIdsSelector.mockImplementation(() => ['300']);
    showFlash.process({ action, getState }, dispatch, done);
    expect(dispatch).not.toHaveBeenCalled();
    expect(done).toHaveBeenCalled();
  });
});
