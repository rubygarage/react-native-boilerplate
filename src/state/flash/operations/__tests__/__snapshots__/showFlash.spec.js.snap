// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`showFlash has valid attributes 1`] = `
Object {
  "cancelType": undefined,
  "debounce": 0,
  "latest": true,
  "name": undefined,
  "process": [Function],
  "processOptions": Object {
    "dispatchMultiple": true,
  },
  "throttle": 0,
  "transform": undefined,
  "type": "SHOW_FLASH",
  "validate": undefined,
  "warnTimeout": 60000,
}
`;
